src/data_distribution/first_touch_distribution.hpp

```diff
@@ -69,52 +80,116 @@ namespace argo {
 						std::cerr << msg_fetch_homenode_fail << std::endl;
 						throw std::system_error(std::make_error_code(static_cast<std::errc>(errno)), msg_fetch_homenode_fail);
 						exit(EXIT_FAILURE);
 					}
 					return homenode;
 				}
 
+				virtual std::size_t peek_local_offset (char* const ptr) {
+					std::size_t page_info[page_info_size];
+					std::size_t offset = SIZE_MAX;
+					static const std::size_t rank = argo::backend::node_id();
+					static const std::size_t global_null = base_distribution<instance>::total_size + 1;
+					const std::size_t drift = (ptr - base_distribution<instance>::start_address) % granularity;
+					const std::size_t addr = (ptr - base_distribution<instance>::start_address) / granularity * granularity;
+					const std::size_t owners_dir_window_index = page_info_size * (addr / granularity);
+
+					std::unique_lock<std::mutex> def_lock(owners_mutex, std::defer_lock);
+					def_lock.lock();
+					/* handle all the necessary directory actions for the global address */
+					update_dirs(addr, false);
+					/* spin in case the values other than `ownership` have not been reflected to the local window */
+					do {
+						argo::backend::atomic::_load_local_owners_dir(&page_info, page_info_size, rank, owners_dir_window_index);
+					} while (page_info[2] != global_null && page_info[1] == global_null);
+					def_lock.unlock();
+
+					/* Update the homenode if we found one */
+					if(page_info[1] != global_null) {
+						offset = page_info[1] + drift;
+					}
+					if( offset != SIZE_MAX &&
+						offset >= base_distribution<instance>::size_per_node) {
+						std::cerr << msg_fetch_offset_fail << std::endl;
+						throw std::system_error(std::make_error_code(static_cast<std::errc>(errno)), msg_fetch_offset_fail);
+						exit(EXIT_FAILURE);
+					}
+					return offset;
+				}
+
 				virtual std::size_t local_offset (char* const ptr) {
 					std::size_t offset;
 					static const std::size_t rank = argo::backend::node_id();
 					static const std::size_t global_null = base_distribution<instance>::total_size + 1;
 					const std::size_t drift = (ptr - base_distribution<instance>::start_address) % granularity;
 					const std::size_t addr = (ptr - base_distribution<instance>::start_address) / granularity * granularity;
-					const std::size_t owners_dir_window_index = 3 * (addr / granularity);
+					const std::size_t owners_dir_window_index = page_info_size * (addr / granularity);
 
 					std::unique_lock<std::mutex> def_lock(owners_mutex, std::defer_lock);
 					def_lock.lock();
 					/* handle all the necessary directory actions for the global address */
 					update_dirs(addr);
 					/* spin in case the values other than `ownership` have not been reflected to the local window */
 					do {
-						argo::backend::atomic::_load_local_owners_dir(&offset, rank, owners_dir_window_index+1);
+						argo::backend::atomic::_load_local_owners_dir(&offset, 1, rank, owners_dir_window_index+1);
```

Comment:
magic number 1, applies to other uses of this function as well. I assume this means there is a directory stored on node 1, but I wonder why node 1 is magic and not node 0.